The deep speech demo for OpenVINO

FengYen-Chang opened the pull request:
Hi, following the policy, I change the branch to develop, and enable the deep speech model in model downloader. Please check this, thanks.

eizamaliev commented on models/public/deep-speech-v0.5.0/deep-speech-v0.5.0.md:
```diff
@@ -0,0 +1,13 @@
+# deep-speech-v0.5.0
```
Please, provide description for input and output in format like [here](https://github.com/opencv/open_model_zoo/blob/develop/models/public/mask_rcnn_resnet101_atrous_coco/mask_rcnn_resnet101_atrous_coco.md).

FengYen-Chang replied:
please check the latest commit for the information of deep speech model.

vladimir-dudnik commented:
Jenkins please retry a build

vladimir-dudnik commented:
@FengYen-Chang you may take a look on [PR](https://github.com/opencv/open_model_zoo/pull/433) for Open Model Zoo Contribution Guide, to check necessary steps.

eizamaliev commented on models/public/deep-speech-v0.5.0/deep-speech-v0.5.0.md:
```diff
@@ -0,0 +1,58 @@
+    - C - number of context
+    - I - number of input
+
+    * Note: Needs to acquire the mfcc feature for audio.
```
```suggestion
    > Note: Needs to acquire the mfcc feature for audio.
```
Also good idea give more explanation, for example add link to demo.

eizamaliev commented on models/public/deep-speech-v0.5.0/deep-speech-v0.5.0.md:
```diff
@@ -0,0 +1,58 @@
+
+## Legal Information
+
+[https://raw.githubusercontent.com/tensorflow/models/master/LICENSE]()
```
I think real lisense is https://raw.githubusercontent.com/mozilla/DeepSpeech/v0.5.0/LICENSE.

FengYen-Chang replied:
Thanks a lot :)

eizamaliev commented on models/public/deep-speech-v0.5.0/deep-speech-v0.5.0.md:
```diff
@@ -0,0 +1,58 @@
+
+### Original Model
+
+The output contain predicted sentence for audio(speech) which decode from ctc beam search decoder. The model was trained on the LibriSpeech clean test corpus for 16-bit, 16 kHz, mono-channel WAVE audio files.
```
Please provide more information, such as name and shape.

eizamaliev commented on models/public/deep-speech-v0.5.0/deep-speech-v0.5.0.md:
```diff
@@ -0,0 +1,58 @@
+
+### Original Model
+
+1. Audio, name: `input_samples`, a WAVE file.
```
There are three input layer in `model.yml` : `input_node`, `previous_state_h/read`, `previous_state_c/read`. Please specify all of them.

(also three outputs)

FengYen-Chang replied:
Thanks for your review. I have updated the file, please check it again, thanks.

eizamaliev commented on models/public/deep-speech-v0.5.0/deep-speech-v0.5.0.md:
```diff
@@ -0,0 +1,97 @@
+
+### Original Model
+
+Output text, name: `Softmax`, Contains predicted sentence for audio(speech) which decode from ctc beam search decoder. The model was trained on the LibriSpeech clean test corpus for 16-bit, 16 kHz, mono-channel WAVE audio files.
```
Please, make the output layers numbered list, like in `Input` section.

FengYen-Chang replied:
Thanks for your review :)

FengYen-Chang commented:
#419

FengYen-Chang commented:
Hi @vladimir-dudnik, I have implemented the accuracy checker for deep speech, could you help me to check this?

vladimir-dudnik commented:
Jenkins please retry a build

vladimir-dudnik commented:
Jenkins please retry a build

vladimir-dudnik commented:
Jenkins please retry a build

vladimir-dudnik commented:
Jenkins please retry a build

vladimir-dudnik commented:
Jenkins please retry a build

vladimir-dudnik commented:
Jenkins please retry a build

vladimir-dudnik commented:
Jenkins please retry a build

vladimir-dudnik commented:
@FengYen-Chang when I try to run demo with 1272-128104-000.wav file error message says that there is no Softmax key. If I change 'Softmax' in deepspeech_openvino_0.5.py, line 178 to 'raw_logits' (what seems to be an output blob) then demo run with runtime warnings but recognize the phrase. Could you please fix runtime warnings?

[ INFO ] Creating Inference Engine
[ INFO ] Loading network files:
	/home/zvdudnik/dev/data/OMZ/PR/models/public/deep-speech-v0.5.0/FP32/deep-speech-v0.5.0.xml
	/home/zvdudnik/dev/data/OMZ/PR/models/public/deep-speech-v0.5.0/FP32/deep-speech-v0.5.0.bin
[ INFO ] Preparing input blobs
[ INFO ] Preparing output blobs
[ INFO ] Loading model to the plugin
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
[ INFO ] Processing output blob
/home/zvdudnik/dev/data/OMZ/PR/open_model_zoo/demos/python_demos/deep_speech_demo/ctc_beamsearch_decoder.py:45: RuntimeWarning: overflow encountered in double_scalars
  _TpB = _pT['l'][_candidate] * pred[_t][idx_b]
/home/zvdudnik/dev/data/OMZ/PR/open_model_zoo/demos/python_demos/deep_speech_demo/ctc_beamsearch_decoder.py:60: RuntimeWarning: overflow encountered in double_scalars
  _TpNB = v * _pT['l'][_candidate]
/home/zvdudnik/dev/data/OMZ/PR/open_model_zoo/demos/python_demos/deep_speech_demo/ctc_beamsearch_decoder.py:57: RuntimeWarning: overflow encountered in double_scalars
  _TpNB = v * _pB['l'][_candidate]
/home/zvdudnik/dev/data/OMZ/PR/open_model_zoo/demos/python_demos/deep_speech_demo/ctc_beamsearch_decoder.py:44: RuntimeWarning: overflow encountered in double_scalars
  _TpNB = _pNB['l'][_candidate] * pred[_t][_candidate[-1]]
/home/zvdudnik/dev/data/OMZ/PR/open_model_zoo/demos/python_demos/deep_speech_demo/ctc_beamsearch_decoder.py:63: RuntimeWarning: invalid value encountered in double_scalars
  _pT['c'][extand_t] += _TpNB
/home/zvdudnik/dev/data/OMZ/PR/open_model_zoo/demos/python_demos/deep_speech_demo/ctc_beamsearch_decoder.py:64: RuntimeWarning: invalid value encountered in double_scalars
  _pNB['c'][extand_t] += _TpNB
/home/zvdudnik/dev/data/OMZ/PR/open_model_zoo/demos/python_demos/deep_speech_demo/ctc_beamsearch_decoder.py:47: RuntimeWarning: invalid value encountered in double_scalars
  _pNB['c'][_candidate] += _TpNB
/home/zvdudnik/dev/data/OMZ/PR/open_model_zoo/demos/python_demos/deep_speech_demo/ctc_beamsearch_decoder.py:51: RuntimeWarning: invalid value encountered in double_scalars
  _pT['c'][_candidate] = _pNB['c'][_candidate] + _pB['c'][_candidate]

>>>mister cuilter is the ampostle of the middle classes and we are glad to welcome his ghospelh

FengYen-Chang commented:
Hi @vladimir-dudnik, Thanks for your testing, I have changed the node from `Softmax` to `raw_logits`, this error cause by different MO version. Since the MO version is different, the output are different and cause the warnings happed. To resolve this, I have added the softmax function.

vladimir-dudnik commented:
Jenkins please retry a build

FengYen-Chang commented:
Hi @vladimir-dudnik, I found I made an error when converting the model automatically, so I patch the `model.yml` to change the output node from `raw_logits` to `Softmax`.

vladimir-dudnik commented:
Jenkins please retry a build

vladimir-dudnik commented:
Jenkins please retry a built

vladimir-dudnik commented:
Jenkins please retry a build

eizamaliev commented on demos/python_demos/deep_speech_demo/README.md:
````diff
@@ -0,0 +1,65 @@
+--output raw_logits,lstm_fused_cell/GatherNd,lstm_fused_cell/GatherNd_1 \ 
+--disable_nhwc_to_nchw
+```
+
````
... or get it from Open Model Zoo using `downloader.py` and `converter.py`

IRDonch replied:
Not even "or", I would say. If the model is downloadable/convertible with our scripts, the README should just tell the user to use them. No need to duplicate this information.

IRDonch commented on models/public/deep-speech-v0.5.0/model.yml:
```diff
@@ -0,0 +1,41 @@
+
+description: >-
+  The `deep-speech` model is a topology intended to perform speech to text.
+  This model is implemented using the tensorflow\* framework.
```
TensorFlow

IRDonch commented on demos/python_demos/deep_speech_demo/speech_features.py:
```diff
@@ -0,0 +1,199 @@
+#
+# John Feng, 2019/09/06 
+#
+##################################################
```
Please remove these banners and add the standard OMZ copyright banners instead. (You already have the correct banner in the main script.)

IRDonch commented on demos/python_demos/deep_speech_demo/deep_speech_demo.py:
```diff
@@ -0,0 +1,195 @@
+    alphabet = Alphabet(os.path.abspath(args.alphabet))
+    
+    # Speech feature extration
+    _wave = wave.open('1272-128104-0000.wav', 'rb')
```
I think you forgot to use the `-i` option here. Also, please use `with` to ensure the file is closed.

IRDonch commented on demos/python_demos/deep_speech_demo/deep_speech_demo.py:
```diff
@@ -0,0 +1,195 @@
+    audio = np.zeros(_length)
+
+    for i in range(_length):
+        audio[i] = int(struct.unpack("<h",_wave.readframes(1))[0])
```
Reading frames one at a time is inefficient. You could replace this with `audio = np.frombuffer(_wave.readframes(_length), dtype=np.dtype('<h'))`.

FengYen-Chang commented:
Close the PR for some issues.
